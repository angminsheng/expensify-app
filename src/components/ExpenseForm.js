import React, { Component } from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import "react-dates/initialize";
import "react-dates/lib/css/_datepicker.css";

export default class ExpenseForm extends Component {
  state = {
    description: this.props.expenses ? this.props.expenses.description : " ",
    note: this.props.expenses ? this.props.expenses.note : " ",
    amount: this.props.expenses ? this.props.expenses.amount.toString() : "",
    createdAt: this.props.expenses
      ? moment(this.props.expenses.createdAt)
      : moment(),
    calenderFocused: false,
    error: ""
  };

  onDescriptionChange = e => {
    let description = e.target.value;
    this.setState({ description });
  };

  onNoteChange = e => {
    let note = e.target.value;
    this.setState({ note });
  };

  onAmountChange = e => {
    let amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      console.log("true");
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = createdAt => {
    if (createdAt) this.setState(() => ({ createdAt }));
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calenderFocused: focused }));
  };

  onSubmit = e => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState({ error: "Please enter description and amount" });
    } else {
      this.setState({ error: "" });
      this.props.onSubmit({
        description: this.state.description,
        amount: Number(this.state.amount),
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error}
        <form action="" onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            value={this.state.description}
            onChange={this.onDescriptionChange}
            autoFocus
          />
          <input
            type="number"
            placeholder="amount"
            onChange={this.onAmountChange}
            value={this.state.amount}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calenderFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            onChange={this.onNoteChange}
            name=""
            id=""
            cols="30"
            rows="10"
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
          />
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}
