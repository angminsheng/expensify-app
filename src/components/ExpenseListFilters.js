import React from "react";
import { connect } from "react-redux";
import {
  sortByAmount,
  setStartDate,
  setEndDate,
  setTextFilter,
  sortByDate
} from "../actions/filters";
import { DateRangePicker } from "react-dates";
import "react-dates/initialize";
import "react-dates/lib/css/_datepicker.css";
// import selectExpenses from "../selectors/expenses";

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = calendarFocused => {
    this.setState({ calendarFocused });
  };

  onInputChange = e => {
    this.props.setTextFilter(e.target.value);
  };

  onSortChange = e => {
    e.target.value === "date"
      ? this.props.sortByDate()
      : this.props.sortByAmount();
  };

  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={this.onInputChange}
        />
        <select
          name=""
          id=""
          value={this.props.filters.sortBy}
          onChange={this.onSortChange}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>

        <DateRangePicker
          startDateId="MyDatePickerStart"
          endDateId="MyDatePickerEnd"
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates
        />
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    filters: state.filters
  };
};

const mapDispatchToProps = dispatch => ({
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate)),
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ExpenseListFilters);
