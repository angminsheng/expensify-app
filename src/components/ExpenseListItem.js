import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";

const ExpenseListItem = props => {
  return (
    <div>
      <Link to={"/edit/" + props.id}>
        <h2>{props.description} </h2>
      </Link>
      {props.amount}-{props.createdAt}
    </div>
  );
};

export default connect()(ExpenseListItem);
