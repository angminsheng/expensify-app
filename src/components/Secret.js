import React from "react";

export default function Secret() {
  return <div>this is the secret page</div>;
}
