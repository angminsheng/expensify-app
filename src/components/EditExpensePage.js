import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { startEditExpense } from "../actions/expenses";
import { startRemoveExpense } from "../actions/expenses";

export const EditExpensePage = props => {
  const onSubmit = expense => {
    props.startEditExpense(props.expenses.id, expense);
    props.history.push("/");
  };

  const onClick = () => {
    props.startRemoveExpense({ id: props.expenses.id });
    props.history.push("/");
  };

  return (
    <div>
      <ExpenseForm
        expenses={props.expenses}
        onSubmit={expense => onSubmit(expense)}
      />
      <button onClick={onClick}>remove</button>
    </div>
  );
};

const mapStateToProps = (state, props) => ({
  expenses: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = dispatch => {
  return {
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: id => dispatch(startRemoveExpense(id))
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);
