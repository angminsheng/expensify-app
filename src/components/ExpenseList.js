import React, { useEffect } from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";
import ExpenseListFilters from "./ExpenseListFilters";
import { startSetExpenses } from "../actions/expenses";

export const ExpenseList = props => {
  // placing the call here will make the data only called here.
  // useEffect(() => {
  //   props.startSetExpense();
  // }, [props]);
  let element = props.expenses.map(expense => (
    <ExpenseListItem {...expense} key={expense.id} />
  ));

  return (
    <div>
      <ExpenseListFilters />
      {props.expenses.length === 0 ? <p>Loading</p> : element}
    </div>
  );
};

//callback function here tells connect what information the component wants to access.

//before map state to props
// export default connect(state => {
//   return {
//     expenses: state.expenses
//   };
// })(ExpenseList);

const mapStateToProps = state => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

const mapDispatchToProps = dispatch => ({
  startSetExpense: () => dispatch(startSetExpenses())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ExpenseList);
