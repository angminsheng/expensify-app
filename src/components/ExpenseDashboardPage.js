import React from "react";

// import { AdminInfo, AuthInfo } from "../playground/hoc";
import ExpenseList from "./ExpenseList";

export const ExpenseDashboardPage = () => (
  <div>
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;
