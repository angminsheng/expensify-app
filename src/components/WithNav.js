import React from "react";
import { Route, Switch } from "react-router-dom";
import AddExpensePage from "./AddExpensePage";
import EditExpensePage from "./EditExpensePage";
import HelpPage from "./HelpPage";
import NotFoundPage from "./NotFoundPage";
import ExpenseDashboardPage from "./ExpenseDashboardPage";
import Header from "./Header";

export default function WithNav() {
  return (
    <div>
      <Header />
      <Switch>
        <Route path="/" component={ExpenseDashboardPage} exact />
        <Route path="/create" component={AddExpensePage} exact />
        <Route path="/edit/:id" component={EditExpensePage} exact />
        <Route path="/help" component={HelpPage} exact />
        <Route component={NotFoundPage} />
      </Switch>
    </div>
  );
}
