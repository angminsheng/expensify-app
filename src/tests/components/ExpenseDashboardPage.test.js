import React from "react";
import { shallow } from "enzyme";
import { ExpenseDashboardPage } from "../../components/ExpenseDashboardPage";
import toJSON from "enzyme-to-json";

describe("Component: ExpenseDashboardPage", () => {
  test("should render the ExpenseDashboardPage", () => {
    const wrapper = shallow(<ExpenseDashboardPage />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
