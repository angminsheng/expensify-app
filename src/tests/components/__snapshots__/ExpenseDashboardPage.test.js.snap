// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: ExpenseDashboardPage should render the ExpenseDashboardPage 1`] = `
<div>
  <Connect(ExpenseList) />
</div>
`;
